neuralhydrology/modelzoo/transformer.py

```diff
@@ -0,0 +1,183 @@
+                                                    nhead=cfg.transformer_nheads, 
+                                                    dim_feedforward=cfg.transformer_dim_feedforward, 
+                                                    dropout=cfg.transformer_dropout)
+        self.encoder = nn.TransformerEncoder(encoder_layer=encoder_layers, 
+                                             num_layers=cfg.transformer_nlayers,
+                                             norm=None) 
+
+        # head (instead of a decoder)
+        self.dropout = nn.Dropout(p=cfg.output_dropout)
+        self.head = get_head(cfg=cfg, n_in=self._encoder_dim, n_out=self.output_size)
+
+        # init weights and biases 
+        self._reset_parameters()
+
+    def _reset_parameters(self): 
+        initrange = 0.1
+        for layer in self.encoder.layers:
+            layer.linear1.weight.data.uniform_(-initrange, initrange)
+            layer.linear1.bias.data.zero_()
+            layer.linear2.weight.data.uniform_(-initrange, initrange)
+            layer.linear2.bias.data.zero_()
+        self.embedding_net._reset_parameters()
+
+    def forward(self, data: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        """Perform a forward pass on a transformer model without decoder.
+
+        Parameters
+        ----------
+        data : Dict[str, torch.Tensor]
+            Dictionary, containing input features as key-value pairs.
+
+        Returns
+        -------
+        Dict[str, torch.Tensor]
+            Model outputs and intermediate states as a dictionary. 
+                - `y_hat`: model predictions of shape [batch size, sequence length, number of target variables].
+        """
+        # transpose to [seq_length, batch_size, n_features]
+        x_d = data['x_d'].transpose(0, 1)
+
+        # concat all inputs
+        if 'x_s' in data and 'x_one_hot' in data:
+            x_s = data['x_s'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_one_hot = data['x_one_hot'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_d = torch.cat([x_d, x_s, x_one_hot], dim=-1)
+        elif 'x_s' in data:
+            x_s = data['x_s'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_d = torch.cat([x_d, x_s], dim=-1)
+        elif 'x_one_hot' in data:
+            x_one_hot = data['x_one_hot'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_d = torch.cat([x_d, x_one_hot], dim=-1)
+        else:
+            pass
+
+        # embedding
+        x_d = self.embedding_net(x_d) * math.sqrt(self._embedding_dim)        
+        x_d = self.pos_encoder(x_d, self.positional_encoding_type)
+
+        # mask past values
+        if self._mask is None or self._mask.size(0) != len(x_d):
+            mask = (torch.tril(torch.ones(len(x_d), len(x_d))) == 1)
+            mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
+            self._mask = mask.to(x_d.device)
```

Comment:
```suggestion
           mask = torch.triu(x_d.new_full((len(x_d), len(x_d)), fill_value=float('-inf')), diagonal=1)
```
This can be reduce to a single line and `x_d.new_full()` makes sure that `mask` is already on the correct device. Please verify that the resulting mask is identical, but my quick test suggest that yes.